File: server/src/intTest/java/org/tctalent/server/repository/db/LanguageLevelRepositoryIntTest.java
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getLanguageLevel;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedLanguageLevel;

import java.util.List;
import java.util.Locale;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.tctalent.server.model.db.LanguageLevel;
import org.tctalent.server.model.db.Status;
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest;

public class LanguageLevelRepositoryIntTest extends BaseDBIntegrationTest {

  @Autowired
  private LanguageLevelRepository repo;
  private LanguageLevel languageLevel;

  @BeforeEach
  public void setup() {
    assertTrue(isContainerInitialised());
    languageLevel = getSavedLanguageLevel(repo);
  }

  @Test
  public void testFindByStatus() {
    List<LanguageLevel> levels = repo.findByStatus(Status.active);
    assertNotNull(levels);
    assertFalse(levels.isEmpty());
    List<String> names = levels.stream().map(LanguageLevel::getName).toList();
    assertTrue(names.contains(languageLevel.getName()));
  }

  @Test
  public void testFindByStatusFail() {
    LanguageLevel newLevel = getLanguageLevel();
    newLevel.setStatus(Status.inactive);
    repo.save(newLevel);
    assertTrue(newLevel.getId() > 0);
    List<LanguageLevel> savedIndustry = repo.findByStatus(Status.active);
    assertNotNull(savedIndustry);
    assertFalse(savedIndustry.isEmpty());
    List<Long> ids = savedIndustry.stream().map(LanguageLevel::getId).toList();
    assertFalse(ids.contains(newLevel.getId()));
  }

  @Test
  public void testFindByNameIgnoreCase() {
    String name = languageLevel.getName().toUpperCase(Locale.getDefault());
    LanguageLevel i = repo.findByNameIgnoreCase(name);
    assertNotNull(i);
    assertEquals(languageLevel.getName(), i.getName());
  }

  // TODO (this test case tests wrong code - i.e. name and code don't match)
  @Test
  public void findByLevelIgnoreCase() {
    // The code actually finds by level and not deleted.
    LanguageLevel level = repo.findByLevelIgnoreCase(1);
    assertNotNull(level);
  }

  // TODO (this test case tests wrong code - i.e. name and code don't match)
  @Test
  public void findByLevelIgnoreCaseFail() {
    // The code actually finds by level and not deleted. Testing the deleted bit.
    languageLevel.setStatus(Status.deleted);
    repo.save(languageLevel);
    LanguageLevel level = repo.findByLevelIgnoreCase(1);
    assertNull(level);
  }

  @Test
  public void findAllActive() {
    // add a second inactive.
    LanguageLevel l = getLanguageLevel();
    l.setStatus(Status.inactive);
    repo.save(l);
    List<LanguageLevel> levels = repo.findAllActive();
    assertNotNull(levels);
    assertFalse(levels.isEmpty());
    List<Long> ids = levels.stream().map(LanguageLevel::getId).toList();
    assertTrue(ids.contains(languageLevel.getId()));
  }
}

File: server/src/intTest/kotlin/org/tctalent/server/repository/db/LanguageLevelRepositoryIntTest.kt
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db

import java.util.*
import kotlin.test.*
import org.springframework.beans.factory.annotation.Autowired
import org.tctalent.server.model.db.LanguageLevel
import org.tctalent.server.model.db.Status
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest
import org.tctalent.server.repository.db.integrationhelp.getLanguageLevel
import org.tctalent.server.repository.db.integrationhelp.getSavedLanguageLevel

class LanguageLevelRepositoryIntTest : BaseDBIntegrationTest() {
  @Autowired lateinit var repo: LanguageLevelRepository
  private lateinit var languageLevel: LanguageLevel

  @BeforeTest
  fun setup() {
    assertTrue { isContainerInitialized() }
    languageLevel = getSavedLanguageLevel(repo)
  }

  @Test
  fun `test find by status`() {
    val levels = repo.findByStatus(Status.active)
    assertNotNull(levels)
    assertTrue { levels.isNotEmpty() }
    val names = levels.map { it.name }
    assertTrue { names.contains(languageLevel.name) }
  }

  @Test
  fun `test find by status fail`() {
    val newLevel = getLanguageLevel().apply { status = Status.inactive }
    repo.save(newLevel)
    assertTrue { newLevel.id > 0 }
    val savedIndustry = repo.findByStatus(Status.active)
    assertNotNull(savedIndustry)
    assertTrue { savedIndustry.isNotEmpty() }
    val ids = savedIndustry.map { it.id }
    assertFalse { ids.contains(newLevel.id) }
  }

  @Test
  fun `test find by name ignore case`() {
    val name = languageLevel.name.uppercase(Locale.getDefault())
    val i = repo.findByNameIgnoreCase(name)
    assertNotNull(i)
    assertEquals(languageLevel.name, i.name)
  }

  // TODO (this test case tests wrong code - i.e. name and code don't match)
  @Test
  fun `find by level ignore case`() {
    // The code actually finds by level and not deleted.
    val level = repo.findByLevelIgnoreCase(1)
    assertNotNull(level)
  }

  // TODO (this test case tests wrong code - i.e. name and code don't match)
  @Test
  fun `find by level ignore case fail`() {
    // The code actually finds by level and not deleted. Testing the deleted bit.
    repo.save(languageLevel.apply { status = Status.deleted })
    val level = repo.findByLevelIgnoreCase(1)
    assertNull(level)
  }

  @Test
  fun `find all active`() {
    // add a second inactive.
    repo.save(getLanguageLevel().apply { status = Status.inactive })
    val level = repo.findAllActive()
    assertNotNull(level)
    assertTrue(level.isNotEmpty())
    val ids = level.map { it.id }
    assertTrue { ids.contains(languageLevel.id) }
  }
}
